flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalDriver.java

```diff
@@ -0,0 +1,195 @@
+	 * @param retrievalPath Path of the ZooKeeper node which contains the leader information
+	 * @param leaderRetrievalEventHandler handler to notify the leader changes.
+	 */
+	public ZooKeeperLeaderRetrievalDriver(
+			CuratorFramework client,
+			String retrievalPath,
+			LeaderRetrievalEventHandler leaderRetrievalEventHandler) throws Exception {
+		this.client = checkNotNull(client, "CuratorFramework client");
+		this.cache = new NodeCache(client, retrievalPath);
+		this.retrievalPath = checkNotNull(retrievalPath);
+
+		this.leaderRetrievalEventHandler = checkNotNull(leaderRetrievalEventHandler);
+
+		client.getUnhandledErrorListenable().addListener(this);
+		cache.getListenable().addListener(this);
+		cache.start();
+
+		client.getConnectionStateListenable().addListener(connectionStateListener);
+
+		running = true;
+	}
+
+	@Override
+	public void close() throws Exception {
+		if (!running) {
+			return;
+		}
+
+		running = false;
+
+		LOG.info("Closing {}.", this);
+
+		client.getUnhandledErrorListenable().removeListener(this);
+		client.getConnectionStateListenable().removeListener(connectionStateListener);
+
+		try {
+			cache.close();
+		} catch (IOException e) {
+			throw new Exception("Could not properly stop the ZooKeeperLeaderRetrievalDriver.", e);
+		}
+	}
+
+	@Override
+	public void nodeChanged() {
+		retrieveLeaderInformationFromZooKeeper();
+	}
+
+	private void retrieveLeaderInformationFromZooKeeper() {
+		if (running) {
+			try {
+				LOG.debug("Leader node has changed.");
+
+				ChildData childData = cache.getCurrentData();
+
+				String leaderAddress;
+				UUID leaderSessionID;
+
+				if (childData == null) {
+					leaderAddress = null;
+					leaderSessionID = null;
+				} else {
+					byte[] data = childData.getData();
+
+					if (data == null || data.length == 0) {
+						leaderAddress = null;
+						leaderSessionID = null;
+					} else {
+						ByteArrayInputStream bais = new ByteArrayInputStream(data);
+						ObjectInputStream ois = new ObjectInputStream(bais);
+
+						leaderAddress = ois.readUTF();
+						leaderSessionID = (UUID) ois.readObject();
+					}
+				}
+
+				leaderRetrievalEventHandler.notifyLeaderAddress(new LeaderInformation(leaderSessionID, leaderAddress));
+			} catch (Exception e) {
+				leaderRetrievalEventHandler.handleError(new Exception("Could not handle node changed event.", e));
+				ExceptionUtils.checkInterrupted(e);
+			}
+		} else {
+			LOG.debug("Ignoring node change notification since the service has already been stopped.");
+		}
+	}
+
+	private void handleStateChange(ConnectionState newState) {
+		switch (newState) {
+			case CONNECTED:
+				LOG.debug("Connected to ZooKeeper quorum. Leader retrieval can start.");
+				break;
+			case SUSPENDED:
+				LOG.warn("Connection to ZooKeeper suspended. Can no longer retrieve the leader from " +
+					"ZooKeeper.");
+				leaderRetrievalEventHandler.notifyLeaderAddress(LeaderInformation.empty());
```

Comment:
I think we should check whether we are still running before calling this method.

Reply:
Yes. We need to check the `running` state.